.github/scripts/promote_branch.sh

```diff
@@ -0,0 +1,180 @@
+    exit 1
+  fi
+}
+
+check_if_any_commits_in_last_week() {
+  NEW_COMMITS=$(git log --oneline --since="$(date --date="7 days ago" +%Y-%m-%d)" | wc -l)
+  if [ $NEW_COMMITS -ne 0 ] ; then
+    echo "There are commits in staging that are less than a week old. Blocking promotion to promotion"
+    echo "Commits less than a week old: $(git log --oneline --since="$(date --date="7 days ago" +%Y-%m-%d)")"
+    exit 1
+  fi
+}
+
+if [ -z "${TARGET_BRANCH}" ]; then
+  echo -e "Error: missing 'target-branch' argument\n\n"
+  print_help
+  exit 1
+fi
+if [ "${TARGET_BRANCH}" == "staging" ]; then
+  SOURCE_BRANCH="development"
+elif [ "${TARGET_BRANCH}" == "production" ]; then
+  SOURCE_BRANCH="staging"
+else
+  echo "Invalid target-branch. Only 'staging' and 'production' are allowed"
+  print_help
+  exit 1
+fi
+if [ -z "${FORK_OWNER}" ]; then
+  echo -e "Error: missing 'fork-owner' argument\n\n"
+  print_help
+  exit 1
+fi
+if [ -z "${GITHUB_TOKEN}" ]; then
+  echo -e "Error: missing 'GITHUB_TOKEN' environment variable\n\n"
+  print_help
+  exit 1
+fi
+
+# Personal access token with appropriate permissions
+token="${GITHUB_TOKEN}"
+
+UPDATE_BRANCH_NAME="${TARGET_BRANCH}-update-"$(date '+%Y_%m_%d__%H_%M_%S')
+TITLE="chore: promote ${SOURCE_BRANCH} to ${TARGET_BRANCH}"
+
+# Clone the repository
+tmpDir=$(mktemp -d)
+releaseServiceCatalogDir=${tmpDir}/release-service-catalog
+mkdir -p ${releaseServiceCatalogDir}
+
+echo -e "---\nPromoting release-service-catalog ${SOURCE_BRANCH} to ${TARGET_BRANCH}\n---\n"
+
+echo -e "Sync fork with upstream:"
+sync_fork_json=$(curl -s -L \
+  -X POST \
+  -H "Accept: application/vnd.github+json" \
+  -H "Authorization: Bearer ${GITHUB_TOKEN}" \
+  -H "X-GitHub-Api-Version: 2022-11-28" \
+  https://api.github.com/repos/${FORK_OWNER}/release-service-catalog/merge-upstream \
+  -d '{"branch":"'${SOURCE_BRANCH}'"}')
+
+echo $sync_fork_json
+
+git clone "git@github.com:$ORG/$REPO.git" ${releaseServiceCatalogDir}
+cd ${releaseServiceCatalogDir}
+git fetch --all --tags --prune
+git remote add myfork git@github.com:$FORK_OWNER/$REPO.git
+
+# A change cannot go into staging if staging and production differ
+if [[ "${TARGET_BRANCH}" == "staging" && "${FORCE_TO_STAGING}" != "true" ]] ; then
+  git checkout origin/staging
+  check_if_branch_differs production
+fi
+
+# A change cannot go into production if the changes in staging are less than a week old
+if [[ "${TARGET_BRANCH}" == "production" && "${OVERRIDE}" != "true" ]] ; then
+  git checkout origin/staging
+  check_if_any_commits_in_last_week
+fi
+
+# Create a new branch
+git reset --hard HEAD
+git checkout -b "$UPDATE_BRANCH_NAME" origin/"$TARGET_BRANCH"
+
+# Fetch all content from SOURCE_BRANCH
+git diff origin/"$SOURCE_BRANCH" ^HEAD | git apply -
+
+# Update the task git revision values
+grep -rl " ${SOURCE_BRANCH}" pipelines/ | xargs sed -i "s/ ${SOURCE_BRANCH}/ ${TARGET_BRANCH}/g"
+
+git add .
+git commit -s -S -m "${TITLE}"
```

Comment:
Wait, this doesn't seem right. This way we will completely lose all the commit metadata and just have one generic commit per promotion. Was there an issue with the conflicts that you couldn't get over? Or why else not just create a PR from the source branch plus one commit changing the default branch?

Reply:
Yeah, we are losing all the commit metadata and just having one commit per promotion. That was talked about as we just care about the contents. If you just create a PR from the source branch plus one commit changing the default branch, you will have conflicts because the source branch doesn't have the commit that changed the default branch (like development doesn't have the commit changing development -> staging)

Reply:
the promote layers script incorporates the PR info into the description that are being synced. did we lose that here?

Reply:
Yeah, we lost that here. I'm not sure how to duplicate it here. The branches in this repo diverge because of having to update the taskGitRevision default value. Because they diverge, I can't just find the common ancestor commit like the promote layers script does. The only way I can see around this is if we don't have a default taskGitRevision value in the pipelines at all and the operator passes it based on what overlay you are running in. I'm not so sure that is something we want to do though

Reply:
> The only way I can see around this is if we don't have a default taskGitRevision value in the pipelines at all and the operator passes it based on what overlay you are running in. I'm not so sure that is something we want to do though

I thought that's something we want to do in the future, but we just don't have it today. Or do I misremember?

That would really solve our problems, right? But even if we do want to do it, we need a solution for now anyway...

Reply:
Passing it based on overlay is for sure in our plans. I'm just not sure if removing the default from the yaml definitions is in the plans. My solution for now was to have bad history in staging and production. If we do get to a point where there are no changes between the branches, we can always overwrite staging and production to have nice history

Reply:
Yeah, I guess that's a possibility. But I still wasn't sure why we couldn't just use a simple PR. I thought, ok, what if we do cherry-pick for all the commits instead. That's certainly an option as well. But when I started doing it, I got conflicts as well. And then I realized why. 

Two biggest instances were the commits where we removed samples and then another similar one that removed run.yaml:
```
e567ba4 chore: remove pipeline tests/run.yaml files (#326)
4e90730 chore(RHTAPREL-670): remove resource samples/ dirs (#325)
```

The conflict happens before the development branch removes a file, but the file is slightly different in the other branch (staging) - it doesn't want to simply remove a file which has a slightly different version in the other branch. So these are one-off things that you just have to manually resolve.

And then another instance was this:
```
<<<<<<< HEAD
| taskGitRevision | The revision in the taskGitUrl repo to be used | Yes | staging |
=======
| taskGitRevision | The revision in the taskGitUrl repo to be used | Yes | development |

## Changes in 1.0.1
- Fixed bug where pipeline execution didn't wait for verify-enterprise-contract task to succeed
>>>>>>> dedf3c9 (fix: check enterprise contract before pushing to github)
```

As you can see here, git was just confused here, because the changelog addition for 1.0.1 was just too close to the line with the staging/development difference. This will never happen here again, because of your change today which removes this difference from the README files. But something similar could still happen if we were changing something in one of the pipeline yaml files very near the taskGitUrl default definition.

So based on this, I think we should still pursue either cherry-picking, or merging. Probably merging. But we need a way to fix conflicts manually when needed. That means that the script should have an option pick up where you left off when there is a conflict - only do the rest, which is probably just create the PR. WDYT?

Reply:
Yeah, what you are saying makes sense and those are the same merge conflicts I was seeing. The issue though is that one of the goals for this is for it to be automated. The script is manually called now, but we want it to run with a github action or something. So manually resolving conflicts blocks that. If we wanted to do it with a force push instead of PR, it seems a lot easier, or if we wanted to do it semi-manual, it seems a lot easier, but if we actually want to hit both the automated PR goal, I don't know of another way to do it than what I have written here

Reply:
So, if we agree to pass this by the release service when creating the pipeline runs and remove the defaults (or the default could be just one everywhere - either production or development). Then whatever we do here is only temporary, right?

So for this temporary solution, we can either pursue the merge route and accept the fact that will sometimes require manual conflict resolution. Or we can do what you proposed here - completely remove the original commits and make one commit per promotion - with the plan to fix this with a force push from `development` eventually. I guess I'm fine with either, but personally I would lean towards the first option.

Reply:
Yes, hopefully it is only temporary as it would be much nicer to have a clean history instead of all this diverging to maintain `taskGitRevision`. I am meeting with David tomorrow, I will bring this up and ask what his thoughts on it are

Reply:
Resolving this as the branches will all be the same now (and the script will have commit history like the promote overlay script). Feel free to open a new discussion if we want to continue this topic